Firebase/InstanceID/FIRInstanceID.m

```diff
@@ -893,10 +896,19 @@ - (void)fetchDefaultToken {
   [self defaultTokenWithHandler:handler];
 }
 
-- (void)defaultTokenWithHandler:(FIRInstanceIDTokenHandler)handler {
+- (void)defaultTokenWithHandler:(nullable FIRInstanceIDTokenHandler)aHandler {
+  // Store handlers to be called once operation completes
+  if (!self.defaultTokenFetchHandler) {
+    self.defaultTokenFetchHandler = [[FIRInstanceIDCombinedHandler<NSString *> alloc] init];
+  }
+  if (aHandler) {
+    [self.defaultTokenFetchHandler addHandler:aHandler];
+  }
+
   if (self.isFetchingDefaultToken || self.isDefaultTokenFetchScheduled) {
     return;
```

Comment:
Wow I didn't expect to be this much change. I'm simply expecting a handler should be handled right before this return.

Reply:
@chliangGoogle Do you mean we should just call the passed handler here before return? What should we pass there then?

Reply:
Actually the fix is to remove isFetchingDefaultToken, because this is always true as it is triggered by core SDK when getting the default token. And if developer happens to ask default token right after, it will block them. So there is no point having this boolean value. WDYT?

Reply:
@chliangGoogle  As far as I can see the flag`isFetchingDefaultToken` prevents `[self tokenWithAuthorizedEntity:...]` from being called one more time until previous call finished. If we remove it, then we will have `[self tokenWithAuthorizedEntity:...]` running  several (two in the case described in the issue) times. Not sure if it is handled properly, because there was a test checking that only a single request is being performed at the time.
Also, the token request looks like a multistep operation which may be expensive, so it looks reasonable to have `isFetchingDefaultToken ` or something like this.

Maybe I am missing something though.

Reply:
For developer's case, core call it once and then developer calls it once, which is by design. So this flag being there will definitely break the case. (also I'm not sure your solution will fix this because the flag is where it triggered the return) If we want to guard extra operations done by the system, can we just use isDefaultTokenFetchScheduled?

Reply:
@chliangGoogle The behaviour I implemented is following: 
- `-[FIRInstanceID instanceIDWithHandler:]` can be called any amount of times
- only one `[self tokenWithAuthorizedEntity:...]` is executed at a time
- if there is token fetch in progress when `-[FIRInstanceID instanceIDWithHandler:]` is called, then the handler will be stored. All the stored handlers will be called once the fetch operation completes.

I added couple unit tests to verify this behaviour. Also I tested the developer's case - the fix works as expected for me.

Having this implemented we may not worry about multiple calls from the client developers side as well as from other SDK components if needed in the future.

As for the `isDefaultTokenFetchScheduled ` - it is currently used to handle retries, so I am not sure it can be used instead of `isFetchingDefaultToken`.

Reply:
But your implementation still doesn't fix the issue here. Line 909 will return without a handler. Maybe the race condition shifted when you add more logic but I don't see how this will resolve line 909, which is where the original issue happened.

Another thing is we already had a queue of FIRInstanceIDTokenFetchOperation to make sure we capture each handler, so this is repetitive.  Again I'm looking for a simple solution, capture a list of handlers don't seem to resolve the core issue here.

Reply:
Oh I see your solution now, your solution is good, but it depends on a boolean value here which is not always accurate, it is not guarded by mutual exclusion. We have plan to some refactor first to simplify the code base. There are way too many calls to fetch tokens right now. I think once we clean it up, we can replay your combine handlers solutions. But right now I think it adds up more complexity.  ( we also have a FIRInstanceIDTokenFetchOperation.m somehow queue up this fetch operations, which is doing the same thing of your class, which I'm thinking now that if we have a solution like yours that always guarantee only one tokenWithAuthorizedEntity:, then maybe we can get rid of the operation queue, which is less effective)
Feel free to help refactor and let me know what you think.

Right now we can change line 909 to below and I"m okay if one more operation going through

if ((self.isFetchingDefaultToken || self.isDefaultTokenFetchScheduled) && !handler) {
    return;
}

Reply:
I was thinking more of this should resolve the issue.
if ((self.isFetchingDefaultToken || self.isDefaultTokenFetchScheduled) && !handler) {
    return;
}

I understand your solution now, I think that's better but might need more touch up on this. 
a. Should we also guard the read and write of isFetchingDefaultToken and isDefaultTokenFetchScheduled so it's more accurate?
b. We have a FIRInstanceIDTokenFetchOperation that is used to queue up these fetch operations, now that with your solution I think we should probably get rid of FIRInstanceIDTokenFetchOperation queue since there's only going to be one fetch request at a time.
c. we should probably do the same for deleteToken requests, feel free to help refactor any of these in the following PRs.

Reply:
@chliangGoogle Thank you.

a. Sorry would you elaborate a bit more on it, I am not sure I got it
b. yeah, agree. BTW, I wanted to discuss some ideas on the refactoring of the class with you as a next step. It looks like the class can become mach more readable if we leverage  Promises to handle combine async operations, retries, etc. There is a nice Google Objective C implementation of promises (https://github.com/google/promises). I am going to present a proposal once I collect required information on it.

Reply:
a. So when you add handlers to your combine handlers, notice there'a @synchronize to guard the write to avoid race condition? I think we should also do the same for boolean value.
b. Promise hasn't been widely adopted by Firebase style guide, definitely check with @ryanwilson on this.

Reply:
@chliangGoogle 
a. the flags are defined as `atomic`. As far as we use them via properties (`self. isDefaultTokenFetchScheduled`) we should be good from the thread-safety point. 
I had to add `@synchronize` to `FIRInstanceIDCombinedHandler` to protect modifying of the array which it not protected by just a property setter/getter.
b. sure, I will.

Reply:
Oh didn't notice they are atomic!